### PR TITLE
Enhancement: New Rule for checking filenames and new type FILE_INFO

### LINKED ISSUE #820: Ensure filenames using `-` or `_`

### PR DESCRIPTION
**Pull request by OskarStark**
Follows #835

Fixes #820

### CODE REVIEW

File: src/Rule/FilenameUsesUnderscoresOnly.php [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+        return null;
+    }
+
+    public function checkFileInfo(\SplFileInfo $fileInfo): ?string
+    {
+        $filename = u($fileInfo->getFilename());
+
+        if ($filename->containsAny('-')) {
+            return sprintf(
+                'Please don\'t use dashes (-) in filename: %s',
```
Comment (smoench) [inline src/Rule/FilenameUsesUnderscoresOnly.php:50]: The message is misleading. There should be hint what should be used instead

Comment (OskarStark) [reply src/Rule/FilenameUsesUnderscoresOnly.php:50]: Indeed, will fix that

File: src/Rule/FilenameUsesDashesOnly.php [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+use App\Annotations\Rule\ValidExample;
+use App\Value\Lines;
+use App\Value\RuleGroup;
+use function Symfony\Component\String\u;
+
+/**
+ * @Description("Ensures a filename uses only dashes (`-`).")
+ * @InvalidExample("custom_extensions.rst")
+ * @ValidExample("custom-extensions.rst")
+ * @ValidExample("_custom-extensions.rst")
```
Comment (OskarStark) [inline src/Rule/FilenameUsesDashesOnly.php:27]: Looks like the second example is not printed to the rules.md 🤔

File: src/Rule/FilenameUsesDashesOnly.php [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+namespace App\Rule;
+
+use App\Annotations\Rule\Description;
+use App\Annotations\Rule\InvalidExample;
+use App\Annotations\Rule\ValidExample;
+use App\Value\RuleGroup;
+use function Symfony\Component\String\u;
+
+/**
+ * @Description("Ensures a filename uses only dashes (`-`).")
```
Comment (OskarStark) [inline src/Rule/FilenameUsesDashesOnly.php:23]:
```suggestion
 * @Description("Ensures a filename uses only dashes (`-`), but are allowed to start with underscore (`_`). It is a common practice to prefix included files with underscores (`_`).")
```

File: src/Rule/FilenameUsesUnderscoresOnly.php (hunk 1 above)
Comment (OskarStark) [reply src/Rule/FilenameUsesUnderscoresOnly.php]: Done

**Comment by OskarStark** [comment]
Ready to merge from my side

File: src/Rule/FilenameUsesDashesOnly.php [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+         *   _custom-extensions.rst
+         */
+        if ($filename->truncate(1)->equalsTo('_')) {
+            $filename = $filename->slice(1)->toUnicodeString();
+        }
+
+        if ($filename->containsAny('_')) {
+            return sprintf(
+                'Please use dashes (-) for the filename: %s',
+                $fileInfo->getFilename()
```
Comment (alamirault) [inline src/Rule/FilenameUsesDashesOnly.php:52]: With `_foo-bar_baz` message will be `Please use dashes (-) for the filename: _foo-bar_baz`, but user probably want keeping first `_`.

We can maybe suggest expected filename using `u()->camel()`, `_foo-bar-baz`

File: src/Rule/FilenameUsesUnderscoresOnly.php [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+        ];
+    }
+
+    public function check(\SplFileInfo $fileInfo): ?string
+    {
+        $filename = u($fileInfo->getFilename());
+
+        if ($filename->containsAny('-')) {
+            return sprintf(
+                'Please use underscores (_) for the filename: %s',
```
Comment (alamirault) [inline src/Rule/FilenameUsesUnderscoresOnly.php:44]: Suggest expected name ? `u()->snake();`

File: src/Rule/NoPhpOpenTagInCodeBlockPhpDirective.php [hunk 6]
```diff
@@ -36,6 +36,7 @@ public function check(Lines $lines, int $number): ?string
             && !RstParser::codeBlockDirectiveIsTypeOf($line, RstParser::CODE_BLOCK_PHP_ANNOTATIONS, true)
             && !RstParser::codeBlockDirectiveIsTypeOf($line, RstParser::CODE_BLOCK_PHP_ATTRIBUTES, true)
             && !RstParser::codeBlockDirectiveIsTypeOf($line, RstParser::CODE_BLOCK_PHP_SYMFONY, true)
+            && !RstParser::codeBlockDirectiveIsTypeOf($line, RstParser::CODE_BLOCK_PHP_STANDALONE, true)
```
Comment (alamirault) [inline src/Rule/NoPhpOpenTagInCodeBlockPhpDirective.php:39]: Out of scope, but do you have a plan for this condition in many files ?

Comment (OskarStark) [reply src/Rule/NoPhpOpenTagInCodeBlockPhpDirective.php:39]: I tried the DirectiveTrait inPhpCodeBlock method yesterday but it wasn't working 

You can give it a try with this method or introduce a helltet class for this

File: src/Rule/FilenameUsesDashesOnly.php (hunk 4 above)
Comment (OskarStark) [reply src/Rule/FilenameUsesDashesOnly.php:52]: Lets do it in another PR and maybe we can display a real diff like:
```diff
- old-filename
+ new_filename
```

**Pull request merged**